### PR TITLE
Start simplifying collection code in Session

### PR DESCRIPTION
**Pull request by bluetech**
Collection starts in `Session.perform_collect()`. I am trying to understand what it does, but it's very complex (in the cyclomatic complexity sense) and hard to follow. This PR starts to try and reveal the inner beauty of the code :)

The PR is split into commits, each doing one refactoring. Most of them are straightforward (except for `Revert "Move common code between Session and Package to FSCollector"` perhaps) and should be easier to follow than reading the full diff.

The end result is that most of the logic is localized in the `Session.collect()` instead of being spread around many places. The next step would be to figure out what all of the caches and special cases do, and either simplify and document them, or (I suspect) plain remove them if they're not serving a purpose.

### CODE REVIEW

File: src/_pytest/main.py [hunk 1]
```diff
@@ -523,6 +503,43 @@ def gethookproxy(self, fspath: py.path.local):
...
+    ) -> Sequence[nodes.Collector]:
+        assert (
+            path.isfile()
+        ), "{!r} is not a file (isdir={!r}, exists={!r}, islink={!r})".format(
+            path, path.isdir(), path.exists(), path.islink()
+        )
+        ihook = self.gethookproxy(path)
+        if not self.isinitpath(path):
+            if ihook.pytest_ignore_collect(path=path, config=self.config):
+                return ()
```
[the remark this answers is not in the document]
Comment (asottile) [reply src/_pytest/main.py:530]: it's slower and non-idiomatic (the empty tuple object is cached in the code object whereas `tuple()` goes through a full lookup and instantiation)

**Comment by RonnyPfannschmidt** [comment]
@nicoddemus  for refactoring chains like this i believe its good to actually keep them in a chain of commits , as that keeps history more comprehensible

for curated history im against squashing

**Comment by nicoddemus** [comment]
> @nicoddemus for refactoring chains like this i believe its good to actually keep them in a chain of commits , as that keeps history more comprehensible

Good point, agree. 👍

File: src/_pytest/main.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (aklajnert) [reply src/_pytest/main.py:530]: Makes sense, thanks for explaining.

[the remark this answers is not in the document]
Comment (aklajnert) [reply src/_pytest/main.py:530]: Makes sense, thanks for explaining.

File: src/_pytest/main.py [hunk 2]
```diff
@@ -523,6 +503,43 @@ def gethookproxy(self, fspath: py.path.local):
...
+            return False
+        path = py.path.local(direntry.path)
+        ihook = self.gethookproxy(path.dirpath())
+        if ihook.pytest_ignore_collect(path=path, config=self.config):
+            return False
+        norecursepatterns = self.config.getini("norecursedirs")
+        for pat in norecursepatterns:
+            if path.check(fnmatch=pat):
+                return False
+        return True
```
[the remark this answers is not in the document]
Comment (bluetech) [reply src/_pytest/main.py]: Thanks! Applied this and the if-combining suggestion.

[the remark this answers is not in the document]
Comment (aklajnert) [reply src/_pytest/main.py]: I think you can just return negated `any()` instead of having it in the `if` clause:  
```python
return not any(path.check(fnmatch=pat) for pat in norecursepatterns)
```

**Comment by bluetech** [comment]
Thanks everyone who looked at all of these commits :)
